### PR TITLE
Add a benchmarks_monthly.py script to simplify the monthly manual runs

### PR DESCRIPTION
**Pull request by jeffhandley**
We will update the script to capture the build number used for each monthly run. Along with the build number, this script also applies a default filter of '*'. With those two features, the command-line to execute is much simpler.

Instead of: `py(thon3) scripts\benchmarks_ci.py --filter '*' -f net7.0 --dotnet-versions 7.0.100-preview.1.22077.12`
This enables: `py(thon3) scripts\benchmarks_monthly.py net7.0-preview1`

When the run completes, the results are collected into a `.tar.gz` file at the root of the `artifacts` folder. The path to the tar file is logged into the console.

### CODE REVIEW

**Comment by danmoseley** [comment]
Historically I've sometimes needed to pass extra flags to BDN. Should this script allow passing such flags? It's not a big deal as it can easily be edited if necessary.

**Comment by jeffhandley** [comment]
> Historically I've sometimes needed to pass extra flags to BDN. Should this script allow passing such flags? It's not a big deal as it can easily be edited if necessary.

Good catch, @danmoseley. Support for that has been added, along with passing `--architecture`.

File: scripts/benchmarks_monthly.py
Comment (danmoseley) [inline scripts/benchmarks_monthly.py:76]: if run with no arguments, it could emit some example command line(s) perhaps?

File: scripts/benchmarks_monthly.py
Comment (danmoseley) [inline scripts/benchmarks_monthly.py:101]: will this collide if you do a 2nd run on the same day? perhaps include h/m at least eg
```py
datetime.datetime.today().strftime("%Y-%b-%d@%H:%M")
```

File: scripts/benchmarks_monthly.py
Comment (danmoseley) [reply scripts/benchmarks_monthly.py:76]: eh, I guess it's pretty obvious, and most people will be pasting something from an email anyway

**Review by danmoseley — APPROVED** [review]

File: scripts/benchmarks_monthly.py
Comment (jeffhandley) [reply scripts/benchmarks_monthly.py:93]: Uh, yeah... That was... intentional. Uh-huh. Thanks.

**Comment by jeffhandley** [comment]
@adamsitnik I added logic to allow multiple versions to be specified. By default, the sdk and results folders will be cleaned before each run, but a `--no-clean` argument will skip that step. I had started to sort the versions, but that wouldn't have naturally handled `net7.0-rc2` to `net7.0` correctly. So instead, we'll rely on the cleaning doing the trick.

_Also note that with this script, `--filter *` becomes superfluous. :wink:_

@danmoseley and @adamsitnik Can you take another look now that I'm done iterating based on your feedback and ideas? I don't have commit rights, so after I gain re-approval from both of you, I'll need one of you to merge for me please.

**Comment by danmoseley** [comment]
LGTM. I'll let @adamsitnik merge if he signs off.

I could imagine that I might want to rename artifacts/tools instead of delete, if I want to figure a problem out or run again quickly for just a few tests. But it's fine for now -- I can just comment that out, etc.

> Also note that with this script, --filter * becomes superfluous. 😉

Not sure I follow. You still have --filter, also I could imagine it's handy if you want to do this kind of A-B comparison on just a few tests (?)

**Review by danmoseley — APPROVED** [review]
thanks for working on this

**Comment by jeffhandley** [comment]
> > Also note that with this script, --filter * becomes superfluous. 😉
> 
> Not sure I follow. You still have --filter, also I could imagine it's handy if you want to do this kind of A-B comparison on just a few tests (?)

I mean that `--filter '*'` is applied by default, so that specific argument is redundant. But `--filter` is still there if you want to specify one.

File: scripts/benchmarks_monthly.py
Comment (adamsitnik) [inline scripts/benchmarks_monthly.py:97]: I think it would be better to delete everything for given architecture, not just the sdk subfolder. Otherwise, the `dotnet --info` run by python script is still reporting the older versions that were previously downloaded:

```suggestion
    sdkPath = os.path.join(rootPath, 'tools', 'dotnet', args.architecture)
```

Before:

<details>

```log
PS D:\projects\performance> py .\scripts\benchmarks_monthly.py net7.0-preview1 net7.0-preview2 --filter *Burgers* --architecture x86
[2022/03/01 09:51:09][INFO] ----------------------------------------------
[2022/03/01 09:51:09][INFO] Initializing logger 2022-03-01 09:51:09.834451
[2022/03/01 09:51:09][INFO] ----------------------------------------------
[2022/03/01 09:51:09][INFO] rmdir -r .\tools\dotnet\x86\sdk
[2022/03/01 09:51:10][INFO] rmdir -r .\artifacts\bin\MicroBenchmarks\Release\net7.0\BenchmarkDotNet.Artifacts\results
[2022/03/01 09:51:10][INFO] Executing: benchmarks_ci.py --skip-logger-setup --filter *Burgers* --architecture x86 -f net7.0 --dotnet-versions 7.0.100-preview.1.22077.12
[2022/03/01 09:51:10][INFO] Installing tools.
[2022/03/01 09:51:10][INFO] ----------------------
[2022/03/01 09:51:10][INFO] Downloading DotNet Cli
[2022/03/01 09:51:10][INFO] ----------------------
[2022/03/01 09:51:10][INFO] DotNet Install Path: 'D:\projects\performance\tools\dotnet\x86'
[2022/03/01 09:51:10][INFO] Downloading https://dot.net/v1/dotnet-install.ps1
[2022/03/01 09:51:11][INFO] $ pushd "D:\projects\performance"
[2022/03/01 09:51:11][INFO] $ powershell.exe -NoProfile -ExecutionPolicy Bypass D:\projects\performance\tools\dotnet\x86\dotnet-install.ps1 -InstallDir D:\projects\performance\tools\dotnet\x86 -Architecture x86 -Version 7.0.100-preview.1.22077.12
[2022/03/01 09:51:12][INFO] dotnet-install: Note that the intended use of this script is for Continuous Integration (CI) scenarios, where:
[2022/03/01 09:51:12][INFO] dotnet-install: - The SDK needs to be installed without user interaction and without admin rights.
[2022/03/01 09:51:12][INFO] dotnet-install: - The SDK installation doesn't need to persist across multiple CI runs.
[2022/03/01 09:51:12][INFO] dotnet-install: To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.
[2022/03/01 09:51:12][INFO]
[2022/03/01 09:52:04][INFO] dotnet-install: Extracting the archive.
[2022/03/01 09:52:18][INFO] dotnet-install: Adding to current process PATH: "D:\projects\performance\tools\dotnet\x86\". Note: This change will not be visible if PowerShell was run as a child process.
[2022/03/01 09:52:18][INFO] dotnet-install: Note that the script does not resolve dependencies during installation.
[2022/03/01 09:52:18][INFO] dotnet-install: To check the list of dependencies, go to https://docs.microsoft.com/dotnet/core/install/windows#dependencies
[2022/03/01 09:52:18][INFO] dotnet-install: Installation finished
[2022/03/01 09:52:18][INFO] $ popd
[2022/03/01 09:52:18][INFO] $ dotnet --info
[2022/03/01 09:52:18][INFO] .NET SDK (reflecting any global.json):
[2022/03/01 09:52:18][INFO]  Version:   7.0.100-preview.1.22077.12
[2022/03/01 09:52:18][INFO]  Commit:    34e0fba2d3
[2022/03/01 09:52:18][INFO]
[2022/03/01 09:52:18][INFO] Runtime Environment:
[2022/03/01 09:52:18][INFO]  OS Name:     Windows
[2022/03/01 09:52:18][INFO]  OS Version:  10.0.22000
[2022/03/01 09:52:18][INFO]  OS Platform: Windows
[2022/03/01 09:52:18][INFO]  RID:         win10-x86
[2022/03/01 09:52:18][INFO]  Base Path:   D:\projects\performance\tools\dotnet\x86\sdk\7.0.100-preview.1.22077.12\
[2022/03/01 09:52:18][INFO]
[2022/03/01 09:52:18][INFO] Host (useful for support):
[2022/03/01 09:52:18][INFO]   Version: 7.0.0-preview.2.22122.4
[2022/03/01 09:52:18][INFO]   Commit:  1615de6ccc
[2022/03/01 09:52:18][INFO]
[2022/03/01 09:52:18][INFO] .NET SDKs installed:
[2022/03/01 09:52:18][INFO]   7.0.100-preview.1.22077.12 [D:\projects\performance\tools\dotnet\x86\sdk]
[2022/03/01 09:52:18][INFO]
[2022/03/01 09:52:18][INFO] .NET runtimes installed:
[2022/03/01 09:52:18][INFO]   Microsoft.AspNetCore.App 6.0.2 [D:\projects\performance\tools\dotnet\x86\shared\Microsoft.AspNetCore.App]
[2022/03/01 09:52:18][INFO]   Microsoft.AspNetCore.App 7.0.0-preview.1.22077.2 [D:\projects\performance\tools\dotnet\x86\shared\Microsoft.AspNetCore.App]
[2022/03/01 09:52:18][INFO]   Microsoft.AspNetCore.App 7.0.0-preview.2.22107.8 [D:\projects\performance\tools\dotnet\x86\shared\Microsoft.AspNetCore.App]
[2022/03/01 09:52:18][INFO]   Microsoft.AspNetCore.App 7.0.0-preview.3.22117.6 [D:\projects\performance\tools\dotnet\x86\shared\Microsoft.AspNetCore.App]
[2022/03/01 09:52:18][INFO]   Microsoft.NETCore.App 6.0.2 [D:\projects\performance\tools\dotnet\x86\shared\Microsoft.NETCore.App]
[2022/03/01 09:52:18][INFO]   Microsoft.NETCore.App 7.0.0-preview.1.22076.8 [D:\projects\performance\tools\dotnet\x86\shared\Microsoft.NETCore.App]
[2022/03/01 09:52:18][INFO]   Microsoft.NETCore.App 7.0.0-preview.2.22103.2 [D:\projects\performance\tools\dotnet\x86\shared\Microsoft.NETCore.App]
[2022/03/01 09:52:18][INFO]   Microsoft.NETCore.App 7.0.0-preview.2.22122.4 [D:\projects\performance\tools\dotnet\x86\shared\Microsoft.NETCore.App]
[2022/03/01 09:52:18][INFO]   Microsoft.WindowsDesktop.App 6.0.2 [D:\projects\performance\tools\dotnet\x86\shared\Microsoft.WindowsDesktop.App]
[2022/03/01 09:52:18][INFO]   Microsoft.WindowsDesktop.App 7.0.0-preview.1.22076.3 [D:\projects\performance\tools\dotnet\x86\shared\Microsoft.WindowsDesktop.App]
[2022/03/01 09:52:18][INFO]   Microsoft.WindowsDesktop.App 7.0.0-preview.2.22107.1 [D:\projects\performance\tools\dotnet\x86\shared\Microsoft.WindowsDesktop.App]
[2022/03/01 09:52:18][INFO]   Microsoft.WindowsDesktop.App 7.0.0-preview.2.22123.5 [D:\projects\performance\tools\dotnet\x86\shared\Microsoft.WindowsDesktop.App]
```

</details>

After:

<details>

```log
PS D:\projects\performance> py .\scripts\benchmarks_monthly.py net7.0-preview1 net7.0-preview2 --filter *Burgers* --architecture x86
[2022/03/01 13:37:30][INFO] ----------------------------------------------
[2022/03/01 13:37:30][INFO] Initializing logger 2022-03-01 13:37:30.601280
[2022/03/01 13:37:30][INFO] ----------------------------------------------
[2022/03/01 13:37:30][INFO] rmdir -r .\tools\dotnet\x86
[2022/03/01 13:37:31][INFO] rmdir -r .\artifacts\bin\MicroBenchmarks\Release\net7.0\BenchmarkDotNet.Artifacts\results
[2022/03/01 13:37:31][INFO] Executing: benchmarks_ci.py --skip-logger-setup --filter *Burgers* --architecture x86 -f net7.0 --dotnet-versions 7.0.100-preview.1.22077.12
[2022/03/01 13:37:31][INFO] Installing tools.
[2022/03/01 13:37:31][INFO] ----------------------
[2022/03/01 13:37:31][INFO] Downloading DotNet Cli
[2022/03/01 13:37:31][INFO] ----------------------
[2022/03/01 13:37:31][INFO] DotNet Install Path: 'D:\projects\performance\tools\dotnet\x86'
[2022/03/01 13:37:31][INFO] Downloading https://dot.net/v1/dotnet-install.ps1
[2022/03/01 13:37:32][INFO] $ pushd "D:\projects\performance"
[2022/03/01 13:37:32][INFO] $ powershell.exe -NoProfile -ExecutionPolicy Bypass D:\projects\performance\tools\dotnet\x86\dotnet-install.ps1 -InstallDir D:\projects\performance\tools\dotnet\x86 -Architecture x86 -Version 7.0.100-preview.1.22077.12
[2022/03/01 13:37:32][INFO] dotnet-install: Note that the intended use of this script is for Continuous Integration (CI) scenarios, where:
[2022/03/01 13:37:32][INFO] dotnet-install: - The SDK needs to be installed without user interaction and without admin rights.
[2022/03/01 13:37:32][INFO] dotnet-install: - The SDK installation doesn't need to persist across multiple CI runs.
[2022/03/01 13:37:32][INFO] dotnet-install: To set up a development environment or to run apps, use installers rather than this script. Visit https://dotnet.microsoft.com/download to get the installer.
[2022/03/01 13:37:32][INFO]
[2022/03/01 13:38:22][INFO] dotnet-install: Extracting the archive.
[2022/03/01 13:38:36][INFO] dotnet-install: Adding to current process PATH: "D:\projects\performance\tools\dotnet\x86\". Note: This change will not be visible if PowerShell was run as a child process.
[2022/03/01 13:38:36][INFO] dotnet-install: Note that the script does not resolve dependencies during installation.
[2022/03/01 13:38:36][INFO] dotnet-install: To check the list of dependencies, go to https://docs.microsoft.com/dotnet/core/install/windows#dependencies
[2022/03/01 13:38:36][INFO] dotnet-install: Installation finished
[2022/03/01 13:38:36][INFO] $ popd
[2022/03/01 13:38:36][INFO] $ dotnet --info
[2022/03/01 13:38:36][INFO] .NET SDK (reflecting any global.json):
[2022/03/01 13:38:36][INFO]  Version:   7.0.100-preview.1.22077.12
[2022/03/01 13:38:36][INFO]  Commit:    34e0fba2d3
[2022/03/01 13:38:36][INFO]
[2022/03/01 13:38:36][INFO] Runtime Environment:
[2022/03/01 13:38:36][INFO]  OS Name:     Windows
[2022/03/01 13:38:36][INFO]  OS Version:  10.0.22000
[2022/03/01 13:38:36][INFO]  OS Platform: Windows
[2022/03/01 13:38:36][INFO]  RID:         win10-x86
[2022/03/01 13:38:36][INFO]  Base Path:   D:\projects\performance\tools\dotnet\x86\sdk\7.0.100-preview.1.22077.12\
[2022/03/01 13:38:36][INFO]
[2022/03/01 13:38:36][INFO] Host (useful for support):
[2022/03/01 13:38:36][INFO]   Version: 7.0.0-preview.1.22076.8
[2022/03/01 13:38:36][INFO]   Commit:  405337939c
[2022/03/01 13:38:36][INFO]
[2022/03/01 13:38:36][INFO] .NET SDKs installed:
[2022/03/01 13:38:36][INFO]   7.0.100-preview.1.22077.12 [D:\projects\performance\tools\dotnet\x86\sdk]
[2022/03/01 13:38:36][INFO]
[2022/03/01 13:38:36][INFO] .NET runtimes installed:
[2022/03/01 13:38:36][INFO]   Microsoft.AspNetCore.App 7.0.0-preview.1.22077.2 [D:\projects\performance\tools\dotnet\x86\shared\Microsoft.AspNetCore.App]
[2022/03/01 13:38:36][INFO]   Microsoft.NETCore.App 7.0.0-preview.1.22076.8 [D:\projects\performance\tools\dotnet\x86\shared\Microsoft.NETCore.App]
[2022/03/01 13:38:36][INFO]   Microsoft.WindowsDesktop.App 7.0.0-preview.1.22076.3 [D:\projects\performance\tools\dotnet\x86\shared\Microsoft.WindowsDesktop.App]
[2022/03/01 13:38:36][INFO]
```

</details>

**Review by adamsitnik — APPROVED** [review]
this script is going to help us a lot, thank you @jeffhandley!

**Comment by danmoseley** [comment]
It worked over night (on Alpine) -- nice improvement @jeffhandley  thank you!
```
danmose3:~/git/performance# python3 ./scripts/benchmarks_monthly.py net6.0 net7.0-preview2
............
[2022/03/04 17:29:41][INFO] $ popd
[2022/03/04 17:29:41][INFO] Results were created in the following folder:
[2022/03/04 17:29:41][INFO]   ./artifacts/bin/MicroBenchmarks/Release/net7.0/BenchmarkDotNet.Artifacts/results
[2022/03/04 17:29:43][INFO] Results were collected into the following tar archive:
[2022/03/04 17:29:43][INFO]   ./artifacts/2022-03-04-17-29-net7.0-preview2.tar.gz
danmose3:~/git/performance#
danmose3:~/git/performance# ls -ll artifacts/*gz
-rw-r--r--    1 root     root       4932714 Mar  4 08:00 artifacts/2022-03-04-08-00-net6.0.tar.gz
-rw-r--r--    1 root     root       4993423 Mar  4 17:29 artifacts/2022-03-04-17-29-net7.0-preview2.tar.gz
danmose3:~/git/performance# time
```

**Comment by jeffhandley** [comment]
I did notice that when running multiple versions, the first output for results/tar gets lost in the middle of the output. I plan to send a follow-up PR that emits the results output for all versions at the very end.

**Pull request merged**
